### DISCUSSION TITLE
ReferenceError: AbortController is not defined

### DISCUSSION THREAD

**Discussion by sukhcha-in** [vercel/next.js#32736] [Help] [answered]
### What version of Next.js are you using?

^12.0.7

### What version of Node.js are you using?

14.17.3

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

development

### Describe the Bug

I have `Navbar` with search bar which shows suggestions on input change.
`Navbar` is added in `_app.js` so it remains on every page on website.
When we type something into input, I use code below to load suggestions.
```
useEffect(() => {
    const loadData = async () => {
        const res = await fetch('/api/autocomplete?q=' + input)
        const json = await res.json()
        setSuggestions(json.results)
    }
    loadData()
}, [input])
```
API takes few seconds to return data.
When we type something and press enter, it goes to search results page.
But fetch keeps loading even when route is changed.
After few seconds when API returns a response, suggestions list pops up.

Is there any way to cancel all fetch requests when route changes?

I found that `AbortController` might work in this case.
```
const controller = new AbortController()
const signal = controller.signal

useEffect(() => {
    const handleRouteChange = (url, { shallow }) => {
        controller.abort()
    }
    router.events.on('routeChangeStart', handleRouteChange)
}, [])

useEffect(() => {
    const loadData = async () => {
        const res = await fetch('/api/autocomplete?q=' + input, { signal })
        const json = await res.json()
        setSuggestions(json.results)
    }
    loadData()
}, [input])
```
It throws this error:
```
Server Error
ReferenceError: AbortController is not defined
```

### Expected Behavior

I think `AbortController` is only available on client side, but how can I handle this error?

### To Reproduce

Nothing

**Comment by ijjk** [comment, marked as answer]
Hi, you can avoid this error from occurring server-side by moving the `AbortController` initializing inside of your `useEffect` as this is only executed client-side.

**Comment by sukhcha-in** [comment]
@ijjk Yeah it works but `router.events.on('routeChangeStart', handleRouteChange)` gets called multiple times, when input changes. Does this have any issue?
```
useEffect(() => {
    const controller = new AbortController()
    const signal = controller.signal
    const handleRouteChange = (url, { shallow }) => {
        controller.abort()
    }
    router.events.on('routeChangeStart', handleRouteChange)

    const loadData = async () => {
        const res = await fetch('/api/autocomplete?q=' + input, { signal })
        const json = await res.json()
        setSuggestions(json.results)
    }
    loadData()
}, [input])
```

**Reply by ijjk** [reply]
You can return a function for cleanup from `useEffect` and this seems like a better place to handle the abort, see the related documentation here https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect

**Comment by kdeoliveira** [comment]
In case you don't want to move the controller to the client side, you can use [this alternative version](https://github.com/southpolesteve/node-abort-controller) of AbortController for nodejs.
